src/linker/Linker.Steps/RemoveUnreachableBlocksStep.cs

```diff
@@ -102,53 +87,274 @@ void RewriteBodies (Collection<TypeDefinition> types)
+				// Otherwise we would have to remember the inlined code along with the method.
+				//
+				StoreMethodAsProcessedAndRemoveFromQueue (
+					stackNode,
+					AnalyzeMethodForConstantResult (method, reducer.FoldedInstructions));
 			}
 		}
 
-		bool TryInlineBodyDependencies (ref BodyReducer reducer)
+		object AnalyzeMethodForConstantResult (MethodDefinition method, Collection<Instruction> instructions)
 		{
-			bool changed = false;
+			if (!method.HasBody)
+				return NonConstSentinel;
+
+			if (method.ReturnType.MetadataType == MetadataType.Void)
+				return NonConstSentinel;
+
+			switch (Context.Annotations.GetAction (method)) {
+			case MethodAction.ConvertToThrow:
+				return NonConstSentinel;
+			case MethodAction.ConvertToStub:
+				return CodeRewriterStep.CreateConstantResultInstruction (Context, method) ?? NonConstSentinel;
+			}
+
+			if (method.IsIntrinsic () || method.NoInlining)
+				return NonConstSentinel;
+
+			if (!Context.IsOptimizationEnabled (CodeOptimizations.IPConstantPropagation, method))
+				return NonConstSentinel;
+
+			var analyzer = new ConstantExpressionMethodAnalyzer (method, instructions ?? method.Body.Instructions);
+			if (analyzer.Analyze ()) {
+				return analyzer.Result;
+			} else {
+				return NonConstSentinel;
+			}
+		}
+
+		/// <summary>
+		/// Determines if a method has constant return value. If the method has not yet been processed it makes sure
+		/// it is on the stack for processing and returns without a result.
+		/// </summary>
+		/// <param name="method">The method to determine result for</param>
+		/// <param name="constantResultInstruction">If successfull and the method returns a constant value this will be set to the
+		/// instruction with the constant value. If successfulll and the method doesn't have a constant value this is set to null.</param>
+		/// <returns>
+		/// true - if the method was analyzed and result is known
+		///   constantResultInstruction is set to an instance if the method returns a constant, otherwise it's set to null
+		/// false - if the method has not yet been analyzed and the caller should retry later
+		/// </returns>
+		bool TryGetConstantResultForMethod (MethodDefinition method, out Instruction constantResultInstruction)
+		{
+			if (!processedMethods.TryGetValue (method, out object processedState)) {
+				// Method is not yet in the stack - add it there
+				AddMethodForProcessing (method);
+				constantResultInstruction = null;
+				return false;
+			}
+
+			switch (processedState) {
+			case LinkedListNode<ProcessingNode> queueNode:
+				// Method is already in the stack - not yet processed
+				// Move it to the top of the stack
+				processingStack.Remove (queueNode);
+				processingStack.AddFirst (queueNode);
+
+				// Note that stack version is not changing - we're just postponing work, not resolving anything.
+
+				constantResultInstruction = null;
```

Comment:
I don't understand the question. Method can only have one final result. If it's still in the queue (as is this case) it doesn't have a result yet. That's why it returns `false` here - telling the caller that no result is available.

Reply:
true, why do you need the remove/addfirst then?

Reply:
To move it to the top of the stack. This is so that callees are processed before callers next time around. For example if both A->C,B and B->C... after processing A the stack would be B,C,A (first is top). After processing B without the shuffle it would remain B,C,A - and it would get stuck. So the remove/add moves C to the begining so that the stack is C,B,A - C is processed and then B can get processed and so on.

Reply:
Yeah, it looks like this is enough if we don't handle recursion. I couldn't come with a case where this would break ;-)